### PR TITLE
[SR-1571][Sema] Don't synthesize super calls to out-of-band constructors

### PR DESCRIPTION
**Pull request by CodaFi**
#### What's in this pull request?

Lookup can occasionally produce a call to a superclass constructor here that is not in our immediate superclass and we would trust it was completely valid.  Instead, following the idea @slavapestov had, we reject the notion of creating such a constructor entirely unless its decl can prove it is actually in our superclass.  

Thanks to @jrose-apple for the logic behind this patch and for sticking with me during our discussions about this.
#### Resolved bug number: ([SR-1571](https://bugs.swift.org/browse/SR-1571))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

File: lib/Sema/CodeSynthesis.cpp [hunk 1]
```diff
@@ -2048,6 +2048,21 @@ swift::createDesignatedInitOverride(TypeChecker &tc,
   if (superclassCtor->getGenericParams())
     return nullptr;
 
+  // Lookup will sometimes give us initializers that are from the ancestors of
+  // our immediate superclass.  So, from the superclass constructor, we look 
+  // one level up to the enclosing type context which will either be a class 
+  // or an extension.  We can use the type declared in that context to check
+  // if it's our immediate superclass and give up if we didn't.
+  // 
+  // FIXME: Remove this when lookup of initializers becomes restricted to our
```
Comment (slavapestov) [inline lib/Sema/CodeSynthesis.cpp]: This looks good but it would be even better to address the FIXME too -- do you want to file a bug for that?

Comment (CodaFi) [inline lib/Sema/CodeSynthesis.cpp]: Even better, I can provide a patch. @jrose-apple discussed the idea of plumbing `NL_VisitSupertypes` through as a proper part of `NameLookupOptions` and using it to affect `lookupConstructors`.

Comment (slavapestov) [inline lib/Sema/CodeSynthesis.cpp]: Sounds good, thanks!

Although I think in this case, the problem is that piece of code we saw in NameLookup which checks if the found name was a protocol requirement, and looks up the witness -- I think this needs a special case where if the witness is a constructor, we skip it (since we would have found it otherwise I think)?

Comment (CodaFi) [inline lib/Sema/CodeSynthesis.cpp]: But I'd like to do that in a different pull request.  When I tried something like this before I have bad memories of breaking more tests than just this one.

Comment (slavapestov) [inline lib/Sema/CodeSynthesis.cpp]: Alright, sounds good!

When you do the cleanup, I think it will suffice to verify that the existing tests pass.

Comment (CodaFi) [inline lib/Sema/CodeSynthesis.cpp]: > I think this needs a special case where if the witness is a constructor, we skip it (since we would have found it otherwise I think)?

The witness is the initializer in A and we're back to square one.  Or, if I special-case this, I can't see where the super call actually goes, so I'm stuck on top of breaking any initializers in extensions that return `Self` et al. like `RawRepresentable`.

**Comment by slavapestov** [comment]
@swift-ci Please test

File: lib/Sema/CodeSynthesis.cpp (hunk 1 above)
Comment (slavapestov) [inline lib/Sema/CodeSynthesis.cpp]: By 'witness' I mean the implementation of the protocol requirement in the special code path, not the lookup result in general.

Let's discuss this in person tomorrow.

**Comment by CodaFi** [comment]
Linux failure looks unrelated.

File: lib/Sema/CodeSynthesis.cpp (hunk 1 above)
Comment (jrose-apple) [inline lib/Sema/CodeSynthesis.cpp]: My particular concern is whether we expect to find convenience initializers from protocol extensions.

Comment (jrose-apple) [inline lib/Sema/CodeSynthesis.cpp]: We also talked about whether it *ever* makes sense to find initializers that aren't immediately on a type or its immediate protocols, but that's potentially a much more disruptive change.

**Comment by jrose-apple** [comment]
Let's get it clean anyway.

@swift-ci Please test Linux platform

**Comment by CodaFi** [comment]
Still failing.  Looks like it's @bitjammer's?

**Comment by slavapestov** [comment]
I pushed a fix for that just now

**Comment by slavapestov** [comment]
@swift-ci Please test and merge

**Comment by CodaFi** [comment]
Still failing in the same test file.

**Comment by bitjammer** [comment]
It looks like the Test and Merge job doesn't have the fix. Might need to run it again.

**Comment by jrose-apple** [comment]
@swift-ci Please test and merge

**Comment by CodaFi** [comment]
Same place, same failure unfortunately.

**Comment by CodaFi** [comment]
@jrose-apple If it helps, I can rebase onto the new changes and see if that helps.

**Comment by harlanhaskins** [comment]
Don't mind me, just

@swift-ci please test

-ing.

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
I have a feeling this and the problems we've been having with #2854 are github bugs.  Seems like it's seeing the rebase pushes as similar enough to the old changes that it squashes them together and reports the old CI results instead of the new ones.

**Comment by CodaFi** [comment]
[Here is the real linux build](https://ci.swift.org/job/swift-PR-Linux/1978/) for this pull request.

**Comment by jrose-apple** [comment]
All right, everything's passing on master, and this test really shouldn't change anything. I'm going to merge it.

**Comment by jrose-apple** [comment]
In the future, please provide a bit more information in your commit message, so that people don't have to go look up the bug to see what's being fixed. (In this case I copy/pasted your description as the commit message.)

Thanks, Robert!
